openhands/controller/agent_controller.py

```diff
@@ -120,12 +127,84 @@ def __init__(
+        # like the regular agent history, it does not include 'hidden' events nor the default filtered out types (backend events)
+        start_id = self.state.start_id if self.state.start_id != -1 else 0
+        end_id = (
+            self.state.end_id
+            if self.state.end_id != -1
+            else self.event_stream.get_latest_event_id()
+        )
+        self.state.history = list(
+            self.event_stream.get_events(
+                start_id=start_id,
+                end_id=end_id,
+                reverse=False,
+                filter_out_type=self.filter_out,
+                filter_hidden=True,
+            )
+        )
+
+        # unsubscribe from the event stream
         self.event_stream.unsubscribe(EventStreamSubscriber.AGENT_CONTROLLER)
 
     def update_state_before_step(self):
         self.state.iteration += 1
         self.state.local_iteration += 1
 
+        # get the history from the event stream
+        # first define the range of events to fetch
+        start_id = self.state.start_id if self.state.start_id != -1 else 0
+        end_id = (
+            self.state.end_id
+            if self.state.end_id != -1
+            else self.event_stream.get_latest_event_id()
+        )
+
+        # fetch events directly from the event stream
+        # filtering out what an agent history should not include:
+        # - "backend" event types that should not be sent to the agent
+        # - hidden events
+        self.state.history = list(
+            self.event_stream.get_events(
+                start_id=start_id,
+                end_id=end_id,
+                reverse=False,
+                filter_out_type=self.filter_out,
+                filter_hidden=True,
+            )
+        )
+
+        # also, we exclude finished delegates from the parent agent's history:
+        # - do not include events between delegate actions and observations
+        # - include the delegate action and observation themselves
+        if self.state.delegates:
+            for (delegate_start_id, delegate_end_id), (
+                delegate_agent,
+                delegate_task,
+            ) in self.state.delegates.items():
+                # sanity checks
+                if (
+                    delegate_start_id < 0
+                    or delegate_end_id < 1
+                    or delegate_start_id >= delegate_end_id
+                    or delegate_end_id >= len(self.state.history)
+                ):
+                    logger.error(
+                        f'Invalid delegate ids: {delegate_start_id}, {delegate_end_id}. Skipping...'
+                    )
+                    continue
+
+                # exclude delegate events from history
+                self.state.history = [
+                    event
+                    for event in self.state.history
+                    if not (delegate_start_id < event.id < delegate_end_id)
+                ]
```

Comment:
this seems like a bit of a hack. Why do these events end up in the delegator's history to begin with?

Reply:
Technically, they didn't, that could have been just a local var. Regardless, appending them in real time fixes this oddness, of course.